File: 5th Semester/Mobile/UI Native Swift/Controller/ReservationController.swift
//
//  RepositoryController.swift
//  Rental
//
//  Created by Mircea Maierean on 09.11.2024.
//
import Foundation

class ReservationController {
    static func createReservation(
        carId: Int,
        userId: UUID,
        startDate: Date,
        endDate: Date,
        pricePerDay: Int,
        repository: InMemoryReservationRepository
    )
    {
        let totalPrice = pricePerDay * totalDays(startDate: startDate, endDate: endDate)
        let reservation = Reservation(
            carId: carId,
            userId: userId,
            totalPrice: totalPrice,
            startDate: startDate,
            endDate: endDate
        )
        repository.addReservation(reservation)
    }
    
    static func totalDays(startDate: Date, endDate: Date) -> Int {
        let days = Calendar.current.dateComponents([.day], from: startDate, to: endDate).day ?? 0
        return days
    }
    
    static func getReservations(repository: InMemoryReservationRepository) -> [Reservation] { 
        return repository.getAllReservations()
    }
    
    static func deleteReservation(id: UUID, repository: InMemoryReservationRepository) {
        repository.deleteReservation(id: id)
    }
    
    static func getReservation(id: UUID, repository: InMemoryReservationRepository) -> Reservation? {
        return repository.getReservation(id: id)
    }
    
   static func updateReservation(reservation: Reservation, repository: InMemoryReservationRepository) {
       if let index = repository.reservations.firstIndex(where: { $0.id == reservation.id }) {
           repository.reservations[index] = reservation
       }
   }
}

File: 5th Semester/mobile/flutter/rental/lib/controller/reservation_controller.dart
import 'package:rental/repository/reservation_repository.dart';
import '../models/reservation.dart';

class ReservationController {
  static void createReservation({
    required int carId,
    required String userId,
    required DateTime startDate,
    required DateTime endDate,
    required int pricePerDay,
    required ReservationRepository repository,
  }) {
    final int totalPrice =
        pricePerDay * totalDays(startDate: startDate, endDate: endDate);
    final reservation = Reservation(
      carId: carId,
      userId: userId,
      totalPrice: totalPrice,
      startDate: startDate,
      endDate: endDate,
    );

    repository.addReservation(reservation);
  }

  static int totalDays(
      {required DateTime startDate, required DateTime endDate}) {
    return endDate.difference(startDate).inDays;
  }

  static List<Reservation> getReservations(
      {required ReservationRepository repository}) {
    return repository.getAllReservations();
  }

  static void deleteReservation({
    required String id,
    required ReservationRepository repository,
  }) {
    repository.deleteReservation(id);
  }

  static Reservation? getReservation({
    required String id,
    required ReservationRepository repository,
  }) {
    return repository.getReservation(id);
  }

  static void updateReservation({
    required Reservation reservation,
    required ReservationRepository repository,
  }) {
    repository.updateReservation(reservation);
  }
}
